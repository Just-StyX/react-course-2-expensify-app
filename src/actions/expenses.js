import { v1 as uuid } from "uuid";
import database from '../firebase/firebase'
import { ref, set, remove, update, onValue, off, push, onChildRemoved, onChildChanged, onChildAdded } from 'firebase/database'

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt }
        const reference = ref(database, 'expenses')
        return push(reference, expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))})
    }
}

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        const reference = ref(database, `expenses/${id}`)
        return remove(reference).then(() => {
            dispatch(removeExpense({ id }))
        })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id, 
    updates
})

export const setExpense = (expenses) => ({
    type: 'SET_EXPENSE',
    expenses
})

export const startSetExpenses = () => {
    return (dispatch) => {
        const reference = ref(database, 'expenses')
        return onValue(reference, (snapshot) => {
            const expenses = [];

            snapshot.forEach((child) => {
                expenses.push({
                    id: child.key,
                    ...child.val()
                })
            })
            dispatch(setExpense(expenses))
        })
    }
}