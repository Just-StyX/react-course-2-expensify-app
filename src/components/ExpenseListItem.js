import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from 'numeral'

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  return (
      <Link className="lists-item" to={`/edit/${id}`}>
        <div>
          <h3 className="lists-item__title">{description}</h3>
          <span className="lists-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</span>
        </div>
        <h3 className="lists-item__data">{numeral(amount / 100).format('$0,0.00')} </h3>
      </Link>
  );
};

export default ExpenseListItem;
