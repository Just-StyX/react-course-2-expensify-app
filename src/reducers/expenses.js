const ADD_EXPENSE = "ADD_EXPENSE";
const REMOVE_EXPENSE = "REMOVE_EXPENSE";
const EDIT_EXPENSE = "EDIT_EXPENSE";
const SET_EXPENSE = "SET_EXPENSE"

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [...state, action.expense];
    case REMOVE_EXPENSE:
      return state.filter(({ id }) => id !== action.id);
    case EDIT_EXPENSE:
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    case SET_EXPENSE:
      return action.expenses
    default:
      return state;
  }
};

export default expensesReducer;
