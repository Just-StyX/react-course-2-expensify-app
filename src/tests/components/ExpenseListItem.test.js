import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import expenses from '../fixtures/expenses'
import ExpenseListItem from '../../components/ExpenseListItem'

test('should render expense list items', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(toJSON(wrapper)).toMatchSnapshot()
})