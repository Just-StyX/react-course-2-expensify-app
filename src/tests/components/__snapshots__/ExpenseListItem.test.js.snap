// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list items 1`] = `
<Link
  className="lists-item"
  to="/edit/1"
>
  <div>
    <h3
      className="lists-item__title"
    >
      Gum
    </h3>
    <span
      className="lists-item__subtitle"
    >
      January 1st, 1970
    </span>
  </div>
  <h3
    className="lists-item__data"
  >
    $1.95
     
  </h3>
</Link>
`;
